xlive/Blam/Engine/objects/objects.cpp

```diff
@@ -1216,4 +1329,97 @@ static int16 __cdecl internal_object_get_markers_by_string_id(datum object_index
 	}
 
 	return (marker != 0 ? (int16)marker_index : 1);
-}
\ No newline at end of file
+}
+
+#ifdef OBJECT_OVERRIDE_ENABLED
+static s_object_override_data* object_override_data_get(size_t override_index)
+{
+	ASSERT(override_index >= 0 && override_index < k_maximum_object_override_count);
+	return &g_object_override_data[override_index];
+}
+
+static int32 object_find_override_index(datum object_index)
+{
+	const object_datum* object = (object_datum*)object_get_and_verify_type(object_index, NONE);
+	int32 override_index = NONE;
+	
+	if (object->object.flags.test(_object_has_override_bit))
+	{
+		// Find the object override with the same object index
+		for (int32 i = 0; i < k_maximum_object_override_count; ++i)
+		{
+			const s_object_override_data* override = object_override_data_get(i);
+			ASSERT(override_index >= 0 && override_index < k_maximum_object_override_count);
+			if (override[i].valid && override[i].object_index == object_index)
+			{
+				override_index = i;
+				break;
+			}
+		}
+	}
+
+	ASSERT(object->object.flags.test(_object_has_override_bit) == (override_index != NONE));
+	return override_index;
+}
+
+static int32 object_get_override_index(datum object_index)
+{
+	object_datum* object = (object_datum*)object_get_and_verify_type(object_index, NONE);
+	int32 override_index = object_find_override_index(object_index);
+	if (override_index == NONE)
+	{
+		ASSERT(!object->object.flags.test(_object_has_override_bit));
+
+		// Initialize the first object override that isn't valid
+		for (int32 i = 0; i < k_maximum_object_override_count; ++i)
+		{
+			s_object_override_data* override = object_override_data_get(i);
+			if (!override->valid)
+			{
+				override_index = i;
+				override->valid = true;
+				override->object_index = object_index;
+				override->shader_tag_index = NONE;
+				break;
+			}
+		}
+
+		if (override_index == NONE)
+		{
+			s_object_override_data* override = object_override_data_get(0);
+			ASSERT(override->valid);
+			ASSERT(override->object_index != NONE);
+
+			object_header_datum* object_header = (object_header_datum*)datum_try_and_get(object_header_data_get(), g_object_override_data[0].object_index);
+			if (object_header)
+			{
+				if (TEST_BIT(0, object_header->type))
```

Comment:
this will always fail?

Reply:
Berth this doesn't look like a bit test, its only doing `1 << object_header->type`

Reply:
changed to use the object_get_and_verify_type function